hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/multipart/MultipartDownloader.java

```diff
@@ -0,0 +1,103 @@
+package org.apache.hadoop.fs.s3a.multipart;
+
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.List;
+
+public final class MultipartDownloader {
+    private static final Log LOG = LogFactory.getLog(MultipartDownloader.class);
+
+    private final long partSize;
+    private final ListeningExecutorService downloadExecutorService;
+    private final PartDownloader partDownloader;
+    private final long chunkSize;
+    private final long bufferSize;
+
+    public MultipartDownloader(long partSize, ListeningExecutorService downloadExecutorService, PartDownloader partDownloader, long chunkSize, long bufferSize) {
+        this.partSize = partSize;
+        this.downloadExecutorService = downloadExecutorService;
+        this.partDownloader = partDownloader;
+        this.chunkSize = chunkSize;
+        this.bufferSize = bufferSize;
```

Comment:
nit: is it worth preconditions sanity check that `partSize > 0`, not null executor & downloader, and positive chunk & buffer size?

Reply:
Do we normally do notNull checks in hadoop? I added the other preconditions.